x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/accesscontrol/RoleSubsetTests.java

```diff
@@ -0,0 +1,191 @@
+                .indices("index-1-1-1-*", "index-1-1-2-*")
+                .privileges("READ")
+                .query("{ \"match\": { \"category\": \"BRD2\" } }")
+                .build();
+        RoleDescriptor baseRD2 = new RoleDescriptor("base-rd-2", null, new IndicesPrivileges[] { baseRD2IP1 }, null);
+
+        final PlainActionFuture<Role> future1 = new PlainActionFuture<>();
+        final Set<RoleDescriptor> baseDescriptors = Sets.newHashSet(baseRD1, baseRD2);
+        CompositeRolesStore.buildRoleFromDescriptors(baseDescriptors, new FieldPermissionsCache(Settings.EMPTY), null, future1);
+        Role baseRole = future1.actionGet();
+
+        // Child role descriptors
+        IndicesPrivileges rd1IP1 = IndicesPrivileges.builder()
+                .indices("index-1-1-1-*")
+                .privileges("READ")
+                .query("{ \"match\": { \"category\": \"RD1\" } }")
+                .build();
+        RoleDescriptor rd1 = new RoleDescriptor("rd-1", null, new IndicesPrivileges[] { rd1IP1 }, null);
+        IndicesPrivileges rd2IP1 = IndicesPrivileges.builder()
+                .indices("index-1-1-2-*")
+                .privileges("READ")
+                .query("{ \"match\": { \"category\": \"RD2\" } }")
+                .build();
+        RoleDescriptor rd2 = new RoleDescriptor("rd-2", null, new IndicesPrivileges[] { rd2IP1 }, null);
+
+        final PlainActionFuture<Role> future2 = new PlainActionFuture<>();
+        final Set<RoleDescriptor> childDescriptors = Sets.newHashSet(rd1, rd2);
+        CompositeRolesStore.buildRoleFromDescriptors(childDescriptors, new FieldPermissionsCache(Settings.EMPTY), null, future2);
+        Role role = future2.actionGet();
+
+        // Check if it is a subset
+        SubsetResult result = role.isSubsetOf(baseRole);
+        assertThat(result.result(), equalTo(SubsetResult.Result.MAYBE));
+        Set<Set<String>> expected = new HashSet<>();
+        expected.add(Sets.newHashSet("index-1-1-1-*"));
+        expected.add(Sets.newHashSet("index-1-1-2-*"));
+        assertThat(result.setOfIndexNamesForCombiningDLSQueries(), equalTo(expected));
+
+        // This is what I need to do next to update the child role descriptors:
+        Map<Set<String>, BoolQueryBuilder> indexNamePatternsToBoolQueryBuilder = new HashMap<>();
+        for (Set<String> indexNamePattern : result.setOfIndexNamesForCombiningDLSQueries()) {
+            BoolQueryBuilder boolQueryBuilder = QueryBuilders.boolQuery();
+            // Now find the index name patterns from all base descriptors that
+            // match and combine queries
+            for (RoleDescriptor rdbase : baseDescriptors) {
+                for (IndicesPrivileges ip : rdbase.getIndicesPrivileges()) {
+                    if (Operations.subsetOf(Automatons.patterns(indexNamePattern), Automatons.patterns(ip.getIndices()))) {
+                        // Do we want to use filter here? do we need scores?
```

Comment:
I do not think this is completely correct. In general we handle multiple DLS queries as being applied being an `OR`. So in general, you will want to do the following: 
* Create a BoolQuery for the index pattern
* Iterate over the user's existing roles, find DLS queries that apply to pattern. Add them as a `should` clause on the boolean query. Set minimum must match to 1 on the bool query.
* Create a new boolquery, add the previously created bool query as a filter. Then add the DLS queries from the scoped roles (I think these are what you call child descriptors) as a should clause. Set minimum must match to 1 there as well.

The boolean query that you just created will become the new DLS query. Take a look at SecurityIndexSearcherWrapper to see how we apply DLS and I think this will clarify why I saw to do things that way.